### ISSUE TITLE
Accessing the 'input' and 'list' components of Autosuggest for inline styling and Velocity animation?

### ISSUE THREAD

**Issue by PAK90** [moroshko/react-autosuggest#120]
First off, thanks for this plugin. I tried rolling my own autocomplete (with ```input``` and a ```ul``` containing a bunch of ```li```), and it's tough stuff. I'm trying to get a few features from my attempt across to the Autosuggest code, but it's all wrapped neatly inside the <Autosuggest> component so I'm not sure how/if I can access it.

In my implementation, the result list items animated in because I surrounded the list items with a ```<VelocityTransitionGroup>```:
```
<VelocityTransitionGroup enter={animationEnter} leave={animationLeave}>                            
    { cardCollection.map(function(card, i){
        return <li key={i} id={i}>{card.name} {card.tagCost}</li>;
    }) }
</VelocityTransitionGroup>
```
I was also styling the ```input``` with my own custom inline style:
```
var borderStyle = {
    borderImageSlice: 1,
    borderImage: '-webkit-linear-gradient(right, '+borderColour2+' 0%, '+borderColour2+' 40%, '+borderColour1+' 60%, '+borderColour1+' 100%) 1'
};  
<input type="text" style={borderStyle}/>
```
I know I can access the suggestion styles via css, but since the borderStyle is a variable controlled by code in the component, it's not possible to push that into the css. So is there any way to get access to the bits I need?

**Comment by moroshko** [comment]
@PAK90 To style the input using inline styles, you'll have to provide `theme` to Autosuggest:
```xml
<Autosuggest theme={theme} ... />
```
This means that you'll need to translate all your Autosuggest styles from pure CSS to inline styles (and put them in the `theme` prop). See [docs](https://github.com/moroshko/react-autosuggest#theme-optional) for the various keys `theme` could have.

Adding `VelocityTransitionGroup` around suggestion items is not possible at the moment. Looking at the [DOM structure](https://github.com/moroshko/react-autosuggest/blob/master/dom-structure.png), are you basically trying to wrap `suggestionsContainer` with `VelocityTransitionGroup`
```xml
<VelocityTransitionGroup enter={animationEnter} leave={animationLeave}>
  [suggestionsContainer]
    [suggestion][/suggestion]
    [suggestion][/suggestion]
  [/suggestionsContainer]
</VelocityTransitionGroup>
```
?

**Comment by PAK90** [comment]
So I could do something like:
```
var theme = {
    .input { borderStyle }
}
```
where I've previously defined ```borderStyle``` and then I'd pass ```theme``` to Autosuggest?

Yeah, wrapping the whole ```suggestionsContainer``` or the individual list of ```suggestions``` I think would do the trick.

**Comment by moroshko** [comment]
> I tried the input style solution and it works, but only statically; I can't get the input border colour to update.

When exactly do you want to update it? Is it based on user interaction?

Is wrapping the whole `Autosuggest` with `VelocityTransitionGroup` (like in your example above) works for you? (at the moment, there is no way to wrap only the suggestions)

**Comment by PAK90** [comment]
Yes, it's based on user interaction. When the user selects an item from the suggestion list certain properties of that item should be reflected in the colour of the input border.

Wrapping the whole Autosuggest didn't work, unfortunately. Another way to use Velocity transitions is directly on the elements, so is there a way of defining a suggestion's onEnter and onLeave?

**Comment by moroshko** [comment]
Try wrapping a single suggestion in your [`renderSuggestion`](https://github.com/moroshko/react-autosuggest#renderSuggestionProp) function.

Regarding the `theme` issue, could you create a Codepen that demonstrates the issue based on one of [these Codepens](http://codepen.io/collection/DkkYaQ)?

**Comment by PAK90** [comment]
I'll try the wrapping in that function when I get back to my code, but I see what you're getting at and I think/hope it'll work.

I mangled your codepen [to create an example](http://codepen.io/anon/pen/vLVXZe?editors=0110). I'm passing the default input theme to ```theme``` except that the border's colour is a variable, which works when I first load it. I made a button to change that variable's value and when you click the button, nothing changes.

**Comment by moroshko** [comment]
@PAK90 To get the input element, you can use Autosuggest's `ref`. No need to grab it by class name (`react-autosuggest__container`). See [this example](http://codepen.io/moroshko/pen/WryOMP).

Currently, there is no API to pass props to `<li>`s. If you could pass props to `<li>`s, would that solve the problem, or you actually have to wrap the `<li>`s with `VelocityComponent`?
`<li>`s are getting props via `itemProps` [here](https://github.com/moroshko/react-autosuggest/blob/master/src/Autosuggest.js#L301) (`itemProps` is currently not in the public API of Autosuggest). Maybe fork the repo and try to add the props you need there, and see if it works. Feel free to submit a Pull Request if you find an elegant solution.

**Comment by moroshko** [comment]
@PAK90 Just wanted to let you know that the `theme` issue not being updated dynamically was resolved in 3.5.0.

**Comment by souporserious** [comment]
Hello 👋  First, I've been having a wonderful time with this plugin, thank you so much for this! I'm wondering if you would be willing to revisit this issue? Is there some possibility of getting a `renderSuggestions` prop to be able to animate children with something like ReactMotion or CSSTransitionGroup? I'd love to be able animate children individually in/out if possible.
